### PR TITLE
[IMAGING-194] Enable Imaging to read TIFF files that use JPEG compression

### PR DESCRIPTION
**Pull request by gwlucastrig**
This pull request enables Imaging to read TIFF files that use the JPEG compression as defined in "Adobe Photoshop(R) TIFF Technical Notes, March 22, 2002".   This enhancement does not support the obsolete JPEG variation that was in use before 2002.

Additional sample files were added to support JUnit testing.

As you review the code changes, you will notice a large block of code that appears to be deleted starting at line 499 in JpegDecoder.java.   I actually did not delete this code, but wrapped it up in a conditional block.  You can see it in its new position starting at line 569.

### CODE REVIEW

File: src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataInterpreterJpeg.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
+/* --------------------------------------------------------------------
+ * Copyright (C) 2023  Gary W. Lucas.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataInterpreterJpeg.java:2]: Is this copyright really necessary? If you leave it in, I'll have to go check and see if we need a notice file update and whatever other legalese is involved.

**Review by garydgregory — CHANGES_REQUESTED** [review]
Thank you for your PR. Processing will be faster and simpler (for me) if you remove the copyright.

File: src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataInterpreterJpeg.java (hunk 1 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataInterpreterJpeg.java:2]: Sorry.  That copyright has absolutely no business being in the code.  My IDE inserted it and I didn't notice.  Will fix.

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffJpegTest.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+import org.apache.commons.imaging.ImagingTestConstants;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Reads files in the BigTIFF samples folder and compares the
+ * BigTIFF files against the standard "classic" tiff image.
+ */
+public class TiffJpegTest extends TiffBaseTest {
```
[the remark this answers is not in the document]
Comment (gwlucastrig) [reply src/test/java/org/apache/commons/imaging/formats/tiff/TiffJpegTest.java:33]: Not sure I follow what you're driving at in this one.  TiffJpegTest does have the "Test" as suffix.  Did you mean "don't change this one when your rename the methods?"

**Comment by gwlucastrig** [comment]
Gary,

Thanks for your comments which, as usual, were spot on.  I will address them as soon as I can.

I did have a question.  Some of the coverage results were not what I expected.  I was looking for an interface that would show me what lines of code were covered and which lines we not.  Maybe I just missed it.  Is one available? Is that feature open only to committers? Also, I did a search for this topic on the developers mailing list, but didn't find any discussion.

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffJpegTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (garydgregory) [reply src/test/java/org/apache/commons/imaging/formats/tiff/TiffJpegTest.java:33]: My mistake, I was looking at the test fixture files. I suppose the word "Test" in the fixture file could be considered superfluous. I'll leave it up to you as to what the best names are.

File: src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java [hunk 3]
```diff
@@ -452,9 +454,21 @@ public void visitSos(final int marker, final byte[] markerBytes, final byte[] im
...
-                colorModel = new DirectColorModel(24, 0x00ff0000, 0x0000ff00, 0x000000ff);
-                final int[] bandMasks = { 0x00ff0000, 0x0000ff00, 0x000000ff };
-                raster = Raster.createPackedRaster(DataBuffer.TYPE_INT, sofnSegment.width, sofnSegment.height, bandMasks, null);
+                // Special handling for the application-RGB case:  TIFF files with
+                // JPEG compression can support an alpha channel.  This extension
+                // to the JPEG standard is implemented by specifying a color model
+                // with a fourth channel for alpha.
+                if (useTiffRgb) {
+                    colorModel = new DirectColorModel(32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000);
+                    final int[] bandMasks = {0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000};
```
[the remark this answers is not in the document]
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:463]: These do operate as constants, so I will define these masks as static final constants. I wanted to change as little code as possible and maintain the coding style used in the original.  Should I also modify the legacy code?

**Comment by garydgregory** [comment]
Hi @gwlucastrig 
Yes, run: `mvn clean site` and  `open target/site/jacoco/index.html`: The site is generated with a JaCoCo report. There is probably a short-hand Maven invocation to create the coverage report instead of the whole site but I can't recall what it is off the top of my head.

**Comment by garydgregory** [comment]
Also, please rebase on git master, I moved the test data folder under resources and had to adjust some tests.

File: src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (garydgregory) [reply src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:463]: If these constants can be refactored, sure. If that refactoring touches a lot of files, consider doing the clean-up in a new PR to make this one a smaller revert/bisect point if we need that in the future. TY!

**Comment by gwlucastrig** [comment]
I am not particularly adept at git, and I can't figure out how to rebase (the rebase attempt fails and git complains that I have unstaged changes).  So I am going to scrap this pull request and start over.  I have already updated the code with your suggestions and will incorporate them into the next pull request.  I will also move the new test files under the new hierarchy.

**Comment by gwlucastrig** [comment]
Well, the "scrap it and start again" option is still available if we need it, but I attempted to rebase and push the changes your requested up to github.   If it didn't work, let me know.

File: src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java [hunk 4]
```diff
@@ -46,6 +46,10 @@
 import org.apache.commons.imaging.formats.jpeg.segments.SosSegment;
 
 public class JpegDecoder extends BinaryFileParser implements JpegUtils.Visitor {
+
+    private static final int [] BAND_MASK_ARGB = {0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000};
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:50]: OK, you define the constants but do not use them. What am I missing?

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hi @gwlucastrig 
Thank you for your update.
Please see my 1 comment.

File: src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java (hunk 4 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:50]: Oops.  I used them on lines 468 and 472.  But I neglected to remove the old lines on 467 and 471.  Will submit and additional change.

**Pull request merged**